### PR TITLE
server: support resigning ddl owner, use http method ddl/owner/resign

### PR DESCRIPTION
**Pull request by winkyao**
### What problem does this PR solve? 

Support resigning DDL owner, let tidb start a new DDL owner election.

### What is changed and how it works?

Add a new HTTP method `ddl/owner/resign` and this method will call  `concurrency.Election.Resign`. `concurrency.Election.Resign` will delete the leader key `/tidb/ddl/fg/owner/` on the etcd. And the owner of TiDB will exit `watchOwner` function, and TiDB cluster will start a new election by `elec.Campaign`.

### Check List 

Tests 

 - Manual test. 

use ansible to start a new cluster with two tidb instances. Run `curl http://{TiDBIP}:10080/ddl/owner/resign` in non-owner will response `This node is not a ddl owner, can't be resigned.`, and the owner will return `"success!"`.

And use `select tidb_is_ddl_owner`, we can find out the owner of the tidb is transferred to the other instance.

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: server/http_handler.go [hunk 1]
```diff
@@ -303,6 +303,25 @@ func (t *tikvHandlerTool) getAllHistoryDDL() ([]*model.Job, error) {
...
+	s, err := session.CreateSession(t.store.(kv.Storage))
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if s != nil {
+		defer s.Close()
+	}
+
+	dom := domain.GetDomain(s.(sessionctx.Context))
```
Comment (lamxTyler) [inline server/http_handler.go]: We can use `session.GetDomain()`.

Comment (winkyao) [reply server/http_handler.go]: What is the difference? Use session.GetDomain() will look likes:

```
dom, err := session.GetDomain(s.GetStore())
	if err != nil {
		return errors.Trace(err)
	}
```

Comment (lamxTyler) [reply server/http_handler.go]: The difference is that we do not need to create a session.

File: docs/tidb_http_api.md [hunk 2]
```diff
@@ -132,3 +132,11 @@ timezone.*
     curl http://{TiDBIP}:10080/tables/{colID}/{colFlag}/{colLen}?rowBin={val}
     ```
     *Hint: For the column which field type is timezone dependent, e.g. `timestamp`, convert its value to UTC timezone.*
+    
+1. Resign the ddl owner, let tidb start a new ddl owner election.
+
+    ```shell
+    curl http://{TiDBIP}:10080/ddl/owner/resign
```
Comment (coocood) [inline docs/tidb_http_api.md]: This should be a `POST` request.

File: server/http_handler.go [hunk 3]
```diff
@@ -334,6 +353,11 @@ type ddlHistoryJobHandler struct {
 	*tikvHandlerTool
 }
 
+// ddlResignOwnerHandler is the handler for resigning ddl owner.
+type ddlResignOwnerHandler struct {
+	*tikvHandlerTool
```
Comment (coocood) [inline server/http_handler.go]: This is not operated on TiKV.
How about we implement the method on `ddlResignOwnerHandler` directly?

**Comment by winkyao** [comment]
@coocood @lamxTyler PTAL

**Comment by coocood** [comment]
Can we just use `elec != nil` to determine it is the owner?

**Comment by winkyao** [comment]
@coocood use `!isOwner()` is more readable.

**Comment by winkyao** [comment]
@coocood fine, if using `elec != nil` to determine, we can remove the mutex.

File: owner/manager.go [hunk 4]
```diff
@@ -179,6 +185,32 @@ func (m *ownerManager) CampaignOwner(ctx context.Context) error {
...
+}
+
+func (m *ownerManager) toBeOwner(elec *concurrency.Election) {
+	atomic.StorePointer(&m.elec, unsafe.Pointer(elec))
+	m.SetOwner(true)
+}
+
+func (m *ownerManager) retireOwner() {
+	m.SetOwner(false)
+	atomic.StorePointer(&m.elec, nil)
```
Comment (lamxTyler) [inline owner/manager.go]: Could we only use the `elec` to determine if it is the owner? Then we do not need to maintain the `owner` anymore.

Comment (winkyao) [reply owner/manager.go]: There are some test cases use SetOwner(false) and just use elec to determin if it is the owner will not easy to mock in tests.

Comment (coocood) [reply owner/manager.go]: Can we change `SetOwner(isOwner bool)` to `SetOwner(elec unsafe.Pointer)`?
In the `mockManager`, we can use any not nil pointer to represent `elec`.

File: owner/manager.go [hunk 5]
```diff
@@ -179,6 +175,31 @@ func (m *ownerManager) CampaignOwner(ctx context.Context) error {
...
+	if elec == nil {
+		return errors.Errorf("This node is not a ddl owner, can't be resigned.")
+	}
+
+	err := elec.Resign(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Warnf("%s Resign ddl owner success!", m.logPrefix)
```
Comment (coocood) [inline owner/manager.go]: Can we call `RetireOwner` here?

Comment (winkyao) [reply owner/manager.go]: No, this is not the work of ResignOwner.

**Comment by coocood** [comment]
LGTM

File: owner/manager.go [hunk 6]
```diff
@@ -45,12 +46,14 @@ type Manager interface {
 	ID() string
 	// IsOwner returns whether the ownerManager is the owner.
 	IsOwner() bool
-	// SetOwner sets whether the ownerManager is the owner.
-	SetOwner(isOwner bool)
+	// RetireOwner make the manager to be a not owner. It's exported for testing
```
Comment (crazycs520) [inline owner/manager.go]: end with .

File: owner/mock.go [hunk 7]
```diff
@@ -49,13 +49,13 @@ func (m *mockManager) IsOwner() bool {
...
-	if isOwner {
-		atomic.StoreInt32(&m.owner, 1)
-	} else {
-		atomic.StoreInt32(&m.owner, 0)
-	}
+func (m *mockManager) toBeOwner() {
+	atomic.StoreInt32(&m.owner, 1)
+}
+
+// setOwner implements Manager.RetireOwner interface.
```
Comment (zimulala) [inline owner/mock.go]: s/setOwner/RetireOwner

File: server/http_handler.go [hunk 8]
```diff
@@ -713,6 +718,37 @@ func (h ddlHistoryJobHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
...
 }
 
+func (h ddlResignOwnerHandler) resignDDLOwner() error {
+	dom, err := session.GetDomain(h.store)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	ownerMgr := dom.DDL().OwnerManager()
+	err = ownerMgr.ResignOwner(context.Background())
```
Comment (zimulala) [inline server/http_handler.go]: Do we need to add a timeout?

**Comment by winkyao** [comment]
@zimulala PTAL

File: server/http_handler.go (hunk 8 above)
Comment (zimulala) [reply server/http_handler.go]: @winkyao Please answer this question.

Comment (winkyao) [reply server/http_handler.go]: Done, PTAL

**Pull request merged**